src/lib/Microsoft.Health.Fhir.Ingest/Service/NormalizationEventConsumerService.cs

```diff
@@ -0,0 +1,218 @@
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Common.Telemetry;
+using Microsoft.Health.Events.EventConsumers;
+using Microsoft.Health.Events.Model;
+using Microsoft.Health.Events.Telemetry;
+using Microsoft.Health.Fhir.Ingest.Data;
+using Microsoft.Health.Fhir.Ingest.Telemetry;
+using Microsoft.Health.Fhir.Ingest.Template;
+using Microsoft.Health.Logging.Telemetry;
+using Newtonsoft.Json.Linq;
+using Polly;
+
+namespace Microsoft.Health.Fhir.Ingest.Service
+{
+    public class NormalizationEventConsumerService : IEventConsumer
+    {
+        private readonly IConverter<IEventMessage, JObject> _converter;
+        private readonly string _templateDefinition;
+        private readonly ITemplateManager _templateManager;
+        private readonly ITelemetryLogger _logger;
+        private readonly IEnumerableAsyncCollector<IMeasurement> _collector;
+        private readonly AsyncPolicy _retryPolicy;
+        private readonly CollectionTemplateFactory<IContentTemplate, IContentTemplate> _collectionTemplateFactory;
+        private readonly IExceptionTelemetryProcessor _exceptionTelemetryProcessor;
+
+        private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1);
+
+        public NormalizationEventConsumerService(
+            IConverter<IEventMessage, JObject> converter,
+            string templateDefinition,
+            ITemplateManager templateManager,
+            IEnumerableAsyncCollector<IMeasurement> collector,
+            ITelemetryLogger logger,
+            CollectionTemplateFactory<IContentTemplate, IContentTemplate> collectionTemplateFactory,
+            NormalizationExceptionTelemetryProcessor exceptionTelemetryProcessor)
+        {
+            _converter = EnsureArg.IsNotNull(converter, nameof(converter));
+            _templateDefinition = EnsureArg.IsNotNullOrWhiteSpace(templateDefinition, nameof(templateDefinition));
+            _templateManager = EnsureArg.IsNotNull(templateManager, nameof(templateManager));
+            _collector = EnsureArg.IsNotNull(collector, nameof(collector));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _retryPolicy = CreateRetryPolicy(logger);
+            _collectionTemplateFactory = EnsureArg.IsNotNull(collectionTemplateFactory, nameof(collectionTemplateFactory));
+            _exceptionTelemetryProcessor = EnsureArg.IsNotNull(exceptionTelemetryProcessor, nameof(exceptionTelemetryProcessor));
+
+            EventMetrics.SetConnectorOperation(ConnectorOperation.Normalization);
+        }
+
+        private (IContentTemplate template, DateTimeOffset timestamp) NormalizationTemplate { get; set; }
+
+        public async Task ConsumeAsync(IEnumerable<IEventMessage> events)
+        {
+            EnsureArg.IsNotNull(events);
+
+            await _retryPolicy.ExecuteAsync(async () => await ConsumeAsyncImpl(events));
+        }
+
+        private async Task<IContentTemplate> GetNormalizationTemplate()

```

Comment:
Still not 100% happy with this.  Main issue is the code is executed per partition on different threads, hence the need for a semaphore to control concurrency when there isn't any horizontal scaling.  A better approach long term would be to move this to a separate class that is injected that will monitor the file and update the template when there are changes, otherwise 

Not, a semaphoresilm is used over a readwritelockslim because the readwritelockslim doesn't work with async calls used in the function.  State is stored in a threadlocal variable so when the awaited call returns, unless we get the same thread, the lock is lost.

Reply:
I had taken a stab at optimizing this in the past. My approach entailed wrapping the current `TemplateManager` in a `CachingTemplateManager`. And caching the results for X minutes. A cache miss could result in multiple threads (one per partition) attempting to fill the cache. But I felt that is ok as the multiple attempts would fill the cache with the same data. Locking could also be implemented inside of the cache population delegate.

This approach goes from instantly picking up customer changes to forcing a new read at regular intervals. So that change would need to be articulated to customers. But even setting the interval to 60 seconds could save hundreds of network calls to the underlying storage account, which would be very useful.

Code is below
```net
using System;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Options;

namespace Microsoft.Health.Fhir.Ingest.Console.Template
{
	public class CachingTemplateManager : ITemplateManager
	{
		private ITemplateManager _wrappedTemplateManager;
        private IMemoryCache _templateCache;


        public CachingTemplateManager(
            ITemplateManager wrappedTemplateManager,
            IMemoryCache cache)
		{
			_wrappedTemplateManager = wrappedTemplateManager;
            _templateCache = cache;
		}

        public byte[] GetTemplate(string templateName)
        {
            var key = $"{templateName}Bytes";
            return _templateCache.GetOrCreate(key,
                e =>
                {
                    e.SetAbsoluteExpiration(TimeSpan.FromMinutes(1));
                    return _wrappedTemplateManager.GetTemplate(templateName);
                });
        }

        public string GetTemplateAsString(string templateName)
        {
            return _templateCache.GetOrCreate(templateName,
                e =>
                {
                    e.SetAbsoluteExpiration(TimeSpan.FromMinutes(1));
                    return _wrappedTemplateManager.GetTemplateAsString(templateName);
                });
        }
    }
}


```

Reply:
Thanks Rob, great suggestion and something similar to this is what I was thinking we would transition to.  I hadn't consider using a cache but that makes sense.  I was thinking we would have a task that fires every so often (5 minutes?) and checks for changes and updates the templates if needed.

The piece that I wanted to include was generating the full template, i.e. I think the TemplateManager should ultimately build and return the completed IContentTemplate.  Otherwise each thread processing a partition is still building new IContentTemplates in memory, increasing our memory pressure and garbage collection.

Reply:
Agreed, it would make sense for the TemplateManager to build the completed template. We can always do a follow up PR to use caching or your suggested timer based approach. Both options reduce the number of network calls to the storage account as well as building the completed IContentTemplate.

Reply:
I think ultimately, we would want to use something like [Blob Storage Events](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-event-overview) if possible. That way we can eliminate all unnecessary storage calls.

Reply:
Just to make sure I understand the suggestion, it is to not refresh the template during execution at all, rather rely on the provisioning flow to trigger the update?  

That is an approach we can take though it doesn't address how new templates are picked up in OSS deployments.  It is a good idea and a good way for us improve this but we will need to have different configurable strategies we can inject for OSS vs managed service.  We are looking at some improvements for how users can manage their templates, perhaps we can do this as part of that work stream.

For now I, I felt it was a pretty substantial departure from how we refresh templates today, so I wanted to be cautious and preserve the existing behavior for now.